base/src/main/java/io/spine/validate/FieldValue.java

```diff
@@ -212,27 +148,68 @@ public JavaType javaType() {
      * Converts the value to a list.
      *
      * @return the value as a list
+     * @deprecated Use {@link #values()} instead.
      */
-    public final ImmutableList<T> asList() {
+    @Deprecated
+    public final ImmutableList<?> asList() {
         return values;
     }
 
-    public T singleValue() {
+    /**
+     * Creates a stream of all the values of this field.
+     *
+     * <p>If the field is singular, obtains the only value. If the field is {@code repeated},
+     * obtains all the values of the list. If the field is a {@code map}, obtains all the map values
+     * (but not the keys).
+     */
+    public final Stream<?> values() {
+        return values.stream();
+    }
+
+    /**
+     * Creates a stream of non-default values of this field.
+     *
+     * <p>Behaves similarly to {@link #values()} but also filters out default values.
+     *
+     * <p>{@code 0} number values, {@code false} boolean values, {@code 0}-number enum instances,
+     * empty char and byte strings, and empty messages are considered default values.
+     */
+    public final Stream<?> nonDefault() {
+        return values()
+                .filter(val -> !isDefault(val));
+    }
+
+    /**
+     * Obtains the single value of this field.
+     *
+     * <p>If the field is a {@linkplain FieldDeclaration#isCollection() collection}, obtains
+     * the first element. If the fist element is not available, throws
+     * an {@code IllegalStateException}.
+     *
+     * @return a single value of this field
+     */
+    public Object singleValue() {
+        checkState(!values.isEmpty(),
+                   "Unable to get the first element of an empty collection for the field `%s`.",
```

Comment:
As far as I understand, the underlying field may be not a collection at all.

Reply:
If the field is not a collection, the list **must** be non-empty, as by `resolveType(..)`.